src/pymor/algorithms/samdp.py

```diff
@@ -0,0 +1,448 @@
+
+            residues = residues[:, :, idx]
+            poles = poles[idx]
+            rightev = rightev[idx]
+            leftev = leftev[idx]
+            if nr_converged < nwanted:
+                logger.warning('The specified number of poles could not be computed.')
+            break
+
+    return poles, residues, rightev, leftev
+
+
+def twosided_rqi(A, E, x, y, theta, init_res, tol, imagtol, maxiter):
+    """Refine an initial guess for an eigentriplet of the matrix pair (A, E).
+
+    Parameters
+    ----------
+    A
+        The |Operator| A from the LTI system.
+    E
+        The |Operator| E from the LTI system.
+    x
+        Initial guess for right eigenvector of matrix pair (A, E).
+    y
+        Initial guess for left eigenvector of matrix pair (A, E).
+    theta
+        Initial guess for eigenvalue of matrix pair (A, E).
+    init_res
+        Residual of initial guess.
+    tol
+        Convergence tolerance for the Rayleigh qotient iteration.
+    imagtol
+        Relative tolerance for imaginary parts of pairs of complex conjugate eigenvalues.
+    maxiter
+        Maximum number of iteration.
+
+    Returns
+    -------
+    x
+        Refined right eigenvector of matrix pair (A, E).
+    y
+        Refined left eigenvector of matrix pair (A, E).
+    theta
+        Refined eigenvalue of matrix pair (A, E).
+    residual
+        Residual of the computed triplet.
+    """
+
+    i = 0
+    nrq = 1
+    while nrq > tol and i < maxiter:
+        i = i + 1
+        Ex = E.apply(x)
+        Ey = E.apply_adjoint(y)
+        tEmA = theta * E - A
+        x_rqi = tEmA.apply_inverse(Ex)
+        v_rqi = tEmA.apply_inverse_adjoint(Ey)
+
+        x_rqi.scal(1 / x_rqi.norm())
+        v_rqi.scal(1 / v_rqi.norm())
+
+        Ax_rqi = A.apply(x_rqi)
+        Ex_rqi = E.apply(x_rqi)
+
+        x_rq = (v_rqi.dot(Ax_rqi) / v_rqi.dot(Ex_rqi))[0][0]
+        if not np.isfinite(x_rq):
+            x_rqi = x
+            v_rqi = y
+            x_rq = theta + 1e-10
+
+        rqi_res = Ax_rqi - Ex_rqi * x_rq
+        if np.abs(np.imag(x_rq)) / np.abs(x_rq) < imagtol:
+            rx_rqi = np.real(x_rqi)
+            rx_rqi.scal(1 / rx_rqi.norm())
+
+            rres = A.apply(rx_rqi) - E.apply(rx_rqi) * np.real(x_rq)
+            nrr = rres.norm()
+            if nrr < rqi_res.norm():
+                x_rqi = rx_rqi
+                v_rqi = np.real(v_rqi)
+                v_rqi.scal(1 / v_rqi.norm())
+                x_rq = np.real(x_rq)
+                rqi_res = rres
+
+        x = x_rqi
+        y = v_rqi
+        theta = x_rq
+        nrq = rqi_res.norm()
+
+        if not np.isfinite(nrq):
+            nrq = 1
+    if nrq < init_res:
+        return x_rqi, v_rqi, x_rq, nrq
+    else:
+        return x, y, theta, init_res
+
+
+def select_max_eig(H, G, X, V, B, C, which):
+    """Compute poles sorted from largest to smallest residual.
+
+    Parameters
+    ----------
+    H
+        The |Numpy array| H from the samdp algorithm.
```

Comment:
The algorithm name should be uppercase. Similar below.
```suggestion
        The |Numpy array| H from the SAMDP algorithm.
```